### ISSUE TITLE
Statically sized matrices with flexible shapes

### ISSUE THREAD

**Issue by amigalemming** [haskell-numerics/hmatrix#131]
This extends the ideas from #128 and #130. We could unify vectors and matrices to `Array shape a`. If we set `shape` to `Max n` we get plain vectors. If we use the `Square size` shape then we get a square matrix. But we can also use it as a vector and multiply it with a matrix with `Square size` as the column dimension. Other shapes might be `Matrix rows cols` for dense rectangular matrices, `LowerTriang size`, `Diagonal size`, `Symmetric size`, `Banded width size` and so on. The values of the shape type are indices, e.g. `data Square size = Square size size`, that is, `Square (Max n)` has all admissible integer indices of a square matrix as values.

**Comment by albertoruiz** [comment]
Ok, I will think about this.

**Comment by stites** [comment]
I'd like to +1 this and mention a different, slightly simpler implementation of this in subhask, which uses type-level strings. See [subhask/example0003-linear-algebra.lhs](https://github.com/mikeizbicki/subhask/blob/master/examples/example0003-linear-algebra.lhs).

**Comment by amigalemming** [comment]
I implemented the idea of a unification of arrays and matrices in two new packages named [comfort-array](https://hackage.haskell.org/package/comfort-array) and [lapack](https://hackage.haskell.org/package/lapack). The first versions actually considered arrays with special matrix shape types as matrices. However, this caused problems. First, it does not work nicely with instances of some classes. E.g. I wanted a `Semigroup` instance for matrix multiplication, but this would mean to implement a `Semigroup` instance for arrays. Second, some matrix types have constraints on their elements. E.g. a Hermitian matrix must have a real diagonal, a unit triangular matrix must have only ones at the diagonal, a band matrix must have zero elements outside the matrix box.
Thus I decided to define Matrix as wrapper around Array with light conversions between them. E.g. the triangular array is allowed to have non-unit elements, but when you convert it to a unit triangular matrix it is checked whether all diagonal elements are one.
